criu/clone-noasan.c

```diff
@@ -27,14 +29,80 @@
  *	clone_noasan
  *	... wait for process to finish ...
  *	unlock_last_pid
+ *
+ * This locking is not required if clone3() is available. lock_last_pid()
+ * and unlock_last_pid() will be noops if clone3() is available.
  */
-int clone_noasan(int (*fn)(void *), int flags, void *arg)
+int clone_noasan(int (*fn)(void *), int flags, void *arg, pid_t pid)
 {
 	void *stack_ptr = (void *)round_down((unsigned long)&stack_ptr - 1024, 16);
+	struct _clone_args c_args = {};
+	long ret;
+
 	BUG_ON((flags & CLONE_VM) && !(flags & CLONE_VFORK));
 	/*
 	 * Reserve some bytes for clone() internal needs
 	 * and use as stack the address above this area.
 	 */
-	return clone(fn, stack_ptr, flags, arg);
+	if (pid == 0)
+		return clone(fn, stack_ptr, flags, arg);
+
+	pr_debug("Creating process using clone3()\n");
+
+	/*
+	 * The simple stack as configured for clone() above those not work
+	 * with clone3(). At the writing of this code it was not clear
+	 * where the difference is. One difference between clone() and
+	 * clone3() is that clone3() wants the stack address and the stack
+	 * size. The kernel does (stack + stack_size) and uses that value
+	 * as the stack pointer for the newly created process.
+	 */
+
+	/*
+	 * The stack_size seems to be pretty big and in the mmap() it is
+	 * even twice the size. By telling the kernel the stack is
+	 * STACK_SIZE and mmap()ing it twice the size the restore actually
+	 * works again. Any different combination did not work. This
+	 * is not understood.
+	 */
```

Comment:
That's a worrying observation. No clue by what this is caused?

Reply:
Right now most of the stack setup, which was working with the old clone implementation (glibc wrapper and own assembler wrapper), is not understood. With the added parameter `stack_size` the whole thing has become more complicated and I do not yet understand why it works and why it crashes.

Reply:
Why did you call the glibc clone but used your own assembler wrapper? That's why glibc clone exists in the first place to not have to do this yourself. 
The stack size was always required even with old clone just implicitly.

Reply:
> Why did you call the glibc clone but used your own assembler wrapper?

I do not understand this. Something is missing in this sentence.

> The stack size was always required even with old clone just implicitly.

It was never a parameter passed to the kernel and it seems the value is not really used in the kernel besides adding it to the stack address.

Reply:
> > Why did you call the glibc clone but used your own assembler wrapper?
> 
> I do not understand this. Something is missing in this sentence.

You said

> Right now most of the stack setup, which was working with the old clone implementation (glibc wrapper and own assembler wrapper), 

That sounds like you used glibc's clone() but still used an assembly wrapper. That sounds strange.

> 
> > The stack size was always required even with old clone just implicitly.
> 
> It was never a parameter passed to the kernel and it seems the value is not really used in the kernel besides adding it to the stack address.

Actually it was for some architectures but that's not the main point. The point is that in both cases the stack_size did matter because of stack direction.

Reply:
Right now I guess I am doing something wrong when setting up the stack in the thread (see criu/pie/restorer.c). I need to look closer at what I might be doing wrong there.

Thanks for all the feedback!

Reply:
If I subtract the stack_size of the new stack pointer for restored threads before calling clone3() the restored process never crashes. Maybe that was already everything that needed to be changed. I will test it some more and then update the PR with my new code. Using Andrei's assembler free clone3() example.